### PR TITLE
Add statx

### PR DESCRIPTION
**Pull request by hasufell**
This is ready for review @hs-viktor @Bodigrim

### CODE REVIEW

File: System/Posix/Files/Common.hsc [hunk 1]
```diff
@@ -688,3 +718,277 @@ getFdPathVar (Fd fd) v =
...
+getExtendedFileStatus_ :: Maybe Fd  -- ^ Optional directory file descriptor
+                       -> CString   -- ^ Pathname to open
+                       -> StatxFlags
+                       -> StatxMask
+                       -> IO ExtendedFileStatus
+#ifdef HAVE_STATX
+getExtendedFileStatus_ fdMay str StatxFlags{..} StatxMask{..} = do
+  fp <- mallocForeignPtrBytes (#const sizeof(struct statx))
+  withForeignPtr fp $ \p ->
+      throwErrnoIfMinus1Retry_ "getExtendedFileStatus_" (c_statx c_fd str flags masks p)
```
Comment (hasufell) [inline System/Posix/Files/Common.hsc:969]: Is `throwErrnoIfMinus1Retry_` good here?

File: System/Posix/Files/Common.hsc [hunk 2]
```diff
@@ -688,3 +718,287 @@ getFdPathVar (Fd fd) v =
...
+-- This is an extended attribute.
+-- Since Linux 5.8.
+fileDaxX                   :: ExtendedFileStatus -> Bool
+
+#if HAVE_STATX
+testFlag :: ExtendedFileStatus -> CAttributes -> Bool
+testFlag ex flag =
+  let attributes = fileAttributesX ex
+      attributes_mask = fileAttributesMaskX ex
+  in (attributes .&. attributes_mask .&. flag) == flag
```
Comment (hasufell) [inline System/Posix/Files/Common.hsc:856]: is my bit code correct?

**Review by vdukhovni — CHANGES_REQUESTED** [review]
Preliminary feedback. Let's make *extensive* use of pattern synonyms to avoid constantly changing the API as code point values come and go.

Tying user-visible API datatypes to a point-in-time set of known values is a strong anti-pattern (no pun intended).

File: System/Posix/Files/Common.hsc [hunk 3]
```diff
@@ -688,3 +718,287 @@ getFdPathVar (Fd fd) v =
...
+    sec  <- (#peek struct statx, stx_atime.tv_sec) statx_ptr :: IO EpochTime
+    nsec <- (#peek struct statx, stx_atime.tv_nsec) statx_ptr :: IO (#type int)
+    let frac = toInteger nsec % 10^(9::Int)
+    return $ fromRational $ toRational sec + frac
+fileCreationTimeHiResX (ExtendedFileStatus statx) =
+  unsafePerformIO $ withForeignPtr statx $ \statx_ptr -> do
+    sec  <- (#peek struct statx, stx_btime.tv_sec) statx_ptr :: IO EpochTime
+    nsec <- (#peek struct statx, stx_btime.tv_nsec) statx_ptr :: IO (#type int)
+    let frac = toInteger nsec % 10^(9::Int)
+    return $ fromRational $ toRational sec + frac
```
Comment (Bodigrim) [inline System/Posix/Files/Common.hsc:907]: `Rational` is likely to be very slow. I'd suggest something like 
```
{-# LANGUAGE NumDecimals #-}
return $ secondsToNominalDiffTime $ MkFixed $ toInteger sec * 1e12 + toInteger nsec * 1e3
```
Could we also please add a test to check that time stamps are interpreted correctly? It's easy to swap `sec` and `nsec`, or treat them as a large `Int64` vs. two `Int32`.

**Review by Bodigrim — COMMENTED** [review]
Thanks!

**Comment by hasufell** [comment]
All done: tests added, using pattern synonyms now, added more missing functionality.

File: tests/FileExtendedStatus.hs [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+ignoreIOExceptions io = io `E.catch`
+                        ((\_ -> return ()) :: IOException -> IO ())
+
+getStatus f = do
+  fs  <- getExtendedFileStatus Nothing f defaultStatxFlags defaultStatxMask
+  ls  <- getExtendedFileStatus Nothing f SymlinkNoFollow defaultStatxMask
+  fs' <- getFileStatus f
+
+  statusExtendedElementsMinimal fs @?= statusElementsMinimal fs'
```
Comment (hasufell) [inline tests/FileExtendedStatus.hs:96]: Here we check the intersection of functionality between `stat` and `statx` is equivalent.

File: tests/FileExtendedStatus.hs [hunk 5]
```diff
@@ -0,0 +1,159 @@
+{-# OPTIONS_GHC -fno-warn-missing-signatures #-}
+
+module FileExtendedStatus (main) where
```
Comment (hasufell) [inline tests/FileExtendedStatus.hs:3]: Mostly a bold copy-pasta of `FileStatus.hs` with some adjustments, but I'm not really bullish of deduplication. Tests are stupid code.

File: System/Posix/Files/Common.hsc [hunk 6]
```diff
@@ -688,3 +752,398 @@ getFdPathVar (Fd fd) v =
...
+pattern StatxAll :: StatxMask
+#ifdef STATX_ALL
+pattern StatxAll = StatxMask (#const STATX_ALL)
+#else
+pattern StatxAll = StatxMask 0
+#endif
+
+
+defaultStatxMask :: StatxMask
+defaultStatxMask = StatxAll
```
Comment (hasufell) [inline System/Posix/Files/Common.hsc:914]: I set the default mask to `STATX_ALL` here.

File: System/Posix/Files/Common.hsc [hunk 7]
```diff
@@ -688,3 +752,398 @@ getFdPathVar (Fd fd) v =
...
+    nsec <- (#peek struct statx, stx_ctime.tv_nsec) statx_ptr :: IO (#type int)
+    return $ timeHiResToNominalDiffTime sec nsec
+fileModificationTimeHiResX (ExtendedFileStatus statx) =
+  unsafePerformIO $ withForeignPtr statx $ \statx_ptr -> do
+    sec  <- (#peek struct statx, stx_mtime.tv_sec) statx_ptr :: IO EpochTime
+    nsec <- (#peek struct statx, stx_mtime.tv_nsec) statx_ptr :: IO (#type int)
+    return $ timeHiResToNominalDiffTime sec nsec
+
+timeHiResToNominalDiffTime :: EpochTime -> Int32 -> POSIXTime
+timeHiResToNominalDiffTime (CTime sec) nsec = secondsToNominalDiffTime $ MkFixed $ toInteger sec * 1e12 + toInteger nsec * 1e3
```
Comment (hasufell) [inline System/Posix/Files/Common.hsc:1081]: This is very juicy.

File: System/Posix/Files/Common.hsc [hunk 8]
```diff
@@ -688,3 +752,398 @@ getFdPathVar (Fd fd) v =
...
+#ifdef STATX_ATTR_DAX
+fileDaxX ex = testFlag ex (#const STATX_ATTR_DAX)
+#else
+{-# WARNING fileDaxX "fileDaxX: not available on this platform" #-}
+fileDaxX _ = error "fileDaxX not available on this platform"
+#endif
+
+fileLinkCountX (ExtendedFileStatus statx) = unsafePerformIO $ do
+  links <- withForeignPtr statx $ (#peek struct statx, stx_nlink)
+  return $ CNlink (fromIntegral (links :: Word32))
```
Comment (hasufell) [inline System/Posix/Files/Common.hsc:1040]: This is not an accident. We get borked results if we use `CNlink` directly:

```
$ cabal run unix-tests -- -p '/fileExtendedStatus/'                                                                               44.8s  Sat 12:45
All
  Native
    fileExtendedStatus: FAIL
      tests/FileExtendedStatus.hs:79:
      expected: 2
       but got: 4294967296002
```

Definitions are:

- statx
  - C code: `__u32`
  - Haskell type: `Word32`
- stat
  - C code: `nlink_t`
  - Haskell type: `CNlink` (wrapped `Word64`)

I'm not too sure what's going on here @vdukhovni.

File: System/Posix/Files/Common.hsc [hunk 9]
```diff
@@ -688,3 +752,398 @@ getFdPathVar (Fd fd) v =
 
 foreign import ccall unsafe "fpathconf"
   c_fpathconf :: CInt -> CInt -> IO CLong
+
+
+-- -----------------------------------------------------------------------------
+-- statx
+--
+
+newtype {-# CTYPE "__u64" #-} CAttributes = CAttributes Int64
```
Comment (hasufell) [inline System/Posix/Files/Common.hsc:761]: Is `__u64` here safe to use?

**Comment by hasufell** [comment]
OpenBSD servers look busted or something

File: System/Posix/Files/Common.hsc [hunk 10]
```diff
@@ -688,3 +753,418 @@ getFdPathVar (Fd fd) v =
...
+    nsec <- (#peek struct statx, stx_ctime.tv_nsec) statx_ptr :: IO (#type int)
+    return $ timeHiResToNominalDiffTime sec nsec
+fileModificationTimeHiResX (ExtendedFileStatus statx) =
+  unsafePerformIO $ withForeignPtr statx $ \statx_ptr -> do
+    sec  <- (#peek struct statx, stx_mtime.tv_sec) statx_ptr :: IO EpochTime
+    nsec <- (#peek struct statx, stx_mtime.tv_nsec) statx_ptr :: IO (#type int)
+    return $ timeHiResToNominalDiffTime sec nsec
+
+timeHiResToNominalDiffTime :: EpochTime -> Int32 -> POSIXTime
+timeHiResToNominalDiffTime (CTime sec) nsec = realToFrac $ picosecondsToDiffTime $ toInteger sec * 1e12 + toInteger nsec * 1e3
```
Comment (hasufell) [inline System/Posix/Files/Common.hsc:1102]: `secondsToNominalDiffTime` is not available on older `time` versions, so have to resort to `picosecondsToDiffTime` and `realToFrac` (which has a rewrite rule between DiffTime and NominalDiffTime)

File: System/Posix/Files/Common.hsc [hunk 11]
```diff
@@ -688,3 +753,443 @@ getFdPathVar (Fd fd) v =
...
+fileCompressedX ex = testFlag ex (#const STATX_ATTR_COMPRESSED)
+#else
+{-# WARNING fileCompressedX "fileCompressedX: not available on this platform (CPP guard: @#if STATX_ATTR_COMPRESSED@)" #-}
+fileCompressedX _ = error "fileCompressedX not available on this platform"
+#endif
+#ifdef STATX_ATTR_IMMUTABLE
+fileImmutableX ex = testFlag ex (#const STATX_ATTR_IMMUTABLE)
+#else
+{-# WARNING fileImmutableX "fileImmutableX: not available on this platform (CPP guard: @#if STATX_ATTR_IMMUTABLE@)" #-}
+fileImmutableX _ = error "fileImmutableX not available on this platform"
```
Comment (vdukhovni) [inline System/Posix/Files/Common.hsc:1049]: Here also, nnd in all similar cases. But the compile-time warnings stay.

File: System/Posix/Files/Common.hsc (hunk 10 above)
Comment (Bodigrim) [reply System/Posix/Files/Common.hsc:1127]: I think it's time we drop support for GHC 8.2 and 8.4, I have not seen anyone using them for ages.

File: System/Posix/Files/Common.hsc (hunk 1 above)
Comment (vdukhovni) [reply System/Posix/Files/Common.hsc:1187]: Are you asking whether it is *necessary* (whether `statx(2)` is interruptible) or whether it is the right thing to do (`statx(2)` is interruptible , but we shouldn't retry)?

The manpage for `statx(2)` does not mention `EINTR`. Plausibly it is not in fact interruptible (caught signals don't interrupt the call).

**Review by vdukhovni — COMMENTED** [review]
Thanks for the `PatternSynonym` rewrite. I hope agree this is the better approach.
Looks good overall, but I haven't meticulously read all the code.
I see some remaining `TODO` comments. Are these for this or future PRs?

**Comment by vdukhovni** [comment]
It seems there are some missing imports. The FreeBSD test CI fails with:
```
System/Posix/Files/Common.hsc:1193:43: error:
    Variable not in scope: ioeSetLocation :: t0 -> String -> IOError
     |
1193 | getExtendedFileStatus_ _ _ _ _ = ioError (ioeSetLocation unsupportedOperation "getExtendedFileStatus")
     |                                           ^^^^^^^^^^^^^^

System/Posix/Files/Common.hsc:1193:58: error:
    Variable not in scope: unsupportedOperation
     |
1193 | getExtendedFileStatus_ _ _ _ _ = ioError (ioeSetLocation unsupportedOperation "getExtendedFileStatus")
     |                                                          ^^^^^^^^^^^^^^^^^^^^
```

**Comment by hasufell** [comment]
> I see some remaining TODO comments. Are these for this or future PRs?

Yeah, it wasn't immediately clear to me how to do them, so they can be added later imo.

File: unix.cabal [hunk 12]
```diff
@@ -69,10 +69,10 @@ library
         buildable: False
 
     build-depends:
-        base        >= 4.10    && < 4.20,
-        bytestring  >= 0.9.2   && < 0.13,
+        base        >= 4.12.0.0  && < 4.20,
```
Comment (hasufell) [inline unix.cabal:72]: Here we restrict to GHC >= 8.6.x

File: unix.cabal [hunk 13]
```diff
@@ -69,10 +69,10 @@ library
...
+        bytestring  >= 0.9.2     && < 0.13,
         filepath    >= 1.4.100.0 && < 1.5,
-        time        >= 1.2     && < 1.13
+        time        >= 1.9.1     && < 1.13
```
Comment (hasufell) [inline unix.cabal:75]: https://hackage.haskell.org/package/time-1.12.2/docs/Data-Time-Clock.html#v:secondsToNominalDiffTime

File: System/Posix/Files/Common.hsc (hunk 10 above)
Comment (hasufell) [reply System/Posix/Files/Common.hsc]: Done

**Review by Bodigrim — APPROVED** [review]
Looks good to me, although I'm no expert in the area.

**Comment by hasufell** [comment]
* made naming more consistent with `fileStatus` variants
* more CPP guards
* added `deviceIDX`, `specialDeviceIDX` and `mountIDX` ...this should be complete now

**Review by vdukhovni — COMMENTED** [review]
Finished read through. Just some cosmetic comments/questions.

The tests are a bit skimpy perhaps. They don't cover much of the fancy `statx` functionality, but admittedly tricky to set up in a portable way (may need root privs or specialised filesystems mounted).  It would be nice to cover some of the more portable features.

That said, basically ready to approve.

**Comment by hasufell** [comment]
* added missing flags
* removed unnecessary CPP
* massively improved docs

Wrt tests... I think I'll leave them as-is for now. Maybe a follow-up PR can add more.

**Review by vdukhovni — APPROVED** [review]
LGTM

**Pull request merged**